sbi/inference/posteriors/vi_posterior.py

```diff
@@ -1,69 +1,481 @@
-    SNLE.<br/><br/>
-    SNLE trains a neural network to approximate the likelihood $p(x|\theta)$. The
-    `SNLE_Posterior` class wraps the trained network such that one can directly evaluate
-    the unnormalized posterior log probability $p(\theta|x) \propto p(x|\theta) \cdot
-    p(\theta)$ and draw samples from the posterior with MCMC.<br/><br/>
-    The neural network itself can be accessed via the `.net` attribute.
+    r"""Provides VI (Variational Inference) to sample from the posterior.<br/><br/>
+
+    SNLE or SNRE train neural networks to approximate the likelihood(-ratios).
+    `VIPosterior` allows to train a tractable variational posterior q(theta) which
+    approximates the true posterior p(theta|x_o). After this second training stage, we
+    can produce approximate posterior samples, by just sampling from q with no
+    additional cost.
     """
 
+    @docstring_parameter(get_default_flows(), get_default_VI_method())
     def __init__(
         self,
         potential_fn: Callable,
-        prior,
+        q: Union[str, Distribution, NeuralPosterior] = "maf",
+        q_kwargs: Dict = dict(),
         theta_transform: Optional[TorchTransform] = None,
-        device: str = "cpu",
+        vi_method: str = "rKL",
+        device: Optional[str] = None,
+        x_shape: Optional[torch.Size] = None,
     ):
         """
         Args:
-            potential_fn:
-            theta_transform:
-            prior: Prior distribution with `.log_prob()` and `.sample()`.
-            proposal: the proposal distribtution (default is the prior).
-            max_sampling_batch_size: the batchsize of samples being drawn from
-                the proposal at every iteration.
-            num_samples_to_find_max: the number of samples that are used to find the
-                maximum of the `potential_fn / proposal` ratio.
-            num_iter_to_find_max: the number of gradient ascent iterations to find the
-                maximum of that ratio.
-            m: multiplier to that ratio.
-            device: Training device, e.g., "cpu", "cuda" or "cuda:{0, 1, ...}".
+            potential_fn: The potential function from which to draw samples.
+            q: Variational family, either string, distribution or an VIPosterior
+                object. This specifies a parameteric class of distribution over which
+                the best possible posterior approximation is searched. For string input
+                we currently support {0}. Of course you can also
+                specify your own variational family by passing a 'parameterized'
+                distribution object i.e. a torch.distributions.Distribution with methods
+                'parameters' returning an iterable of all parameters (you can also
+                handle this using q_kwargs see below).
+            q_kwargs: Arguments for construction q. If q is a string, then this will be
```

Comment:
As I write below, I would completely remove q_kwargs here. If really want to keep them, I would spell them out as individual argument though

Reply:
Yeah, the problem is that every different choice of "q" will also lead to a different choice of "q_kwargs" with some overlap. Thus these kinda add up combinatorially. I did start to write down the main arguments within the specific builders. Not sure if it is enough to refer to them as these are typically not exposed to the 'public'.

Reply:
I would expose the builders, as is also done for e.g. `posterior_nn()`. And then they can have all the arguments spelled out, no?

Reply:
Yes, lets do it like this and remove q_kwargs compeltly.

Reply:
And we change get_flow_builder to return a Callable instead, consistent with likelihood_nn and so on.